src/gas_metering/mod.rs

```diff
@@ -155,185 +166,243 @@ impl Rules for ConstantCostRules {
-				// WebAssembly 1.0 spec, the only allowed element type is funcref.
-				if let GasMeter::External { .. } = gas_meter {
-					for segment in elements_section.entries_mut() {
-						// update all indirect call addresses initial values
-						for func_index in segment.members_mut() {
-							if *func_index >= gas_func_idx {
-								*func_index += 1
-							}
-						}
-					}
+			}
+			if rules.memory_grow_cost().enabled() {
+				let counter;
+				(func_builder, counter) = inject_grow_counter(
+					&FunctionBody::new(0, &truncate_len_from_encoder(&func_builder)?),
+					grow_cnt_func_idx,
+				)?;
+				if counter > 0 {
+					need_grow_counter = true;
 				}
-			},
-			elements::Section::Start(start_idx) =>
-				if let GasMeter::External { .. } = gas_meter {
-					if *start_idx >= gas_func_idx {
-						*start_idx += 1
+			}
+			code_section_builder.function(&func_builder);
+		}
+		module_info.replace_section(SectionId::Code.into(), &code_section_builder)?;
+	}
+
+	if module_info.exports_count > 0 {
+		if let GasMeter::External { .. } = gas_meter {
+			let mut export_sec_builder = ExportSection::new();
+
+			for export in module_info.export_section()?.expect("no export section") {
+				let mut export_index = export.index;
+				if let ExternalKind::Func = export.kind {
+					if export_index >= gas_func_idx {
+						export_index += 1;
 					}
-				},
-			elements::Section::Name(s) =>
-				if let GasMeter::External { .. } = gas_meter {
-					for functions in s.functions_mut() {
-						*functions.names_mut() =
-							IndexMap::from_iter(functions.names().iter().map(|(mut idx, name)| {
-								if idx >= gas_func_idx {
-									idx += 1;
-								}
-
-								(idx, name.clone())
-							}));
+				}
+				export_sec_builder.export(
+					export.name,
+					DefaultTranslator.translate_export_kind(export.kind)?,
+					export_index,
+				);
+			}
+			module_info
+				.replace_section(SectionId::Export.into(), &export_sec_builder)
+				.unwrap();
+		}
+	}
+
+	if module_info.elements_count > 0 {
+		// Note that we do not need to check the element type referenced because in the
+		// WebAssembly 1.0 spec, the only allowed element type is funcref.
+		if let GasMeter::External { .. } = gas_meter {
+			let mut ele_sec_builder = ElementSection::new();
+
+			for elem in module_info
+				.element_section()?
+				.ok_or_else(|| anyhow!("no element_section section"))?
+			{
+				let mut functions = vec![];
+				if let ElementItems::Functions(func_indexes) = elem.items {
+					for func_idx in func_indexes {
+						let mut func_idx = func_idx?;
+						if func_idx >= gas_func_idx {
+							func_idx += 1
+						}
+						functions.push(func_idx);
 					}
-				},
-			_ => {},
+				}
+
+				let offset;
+				//todo edit element is little complex,
+				let mode = match elem.kind {
+					ElementKind::Active { table_index, offset_expr } => {
+						offset = DefaultTranslator.translate_const_expr(
+							&offset_expr,
+							&ValType::I32,
```

Comment:
Why `ValType::I32`? It seems the second argument isn't used anyway. Maybe we just remove it?

Reply:
Makes sense. Will remove it